clang/lib/AST/Type.cpp

```diff
@@ -3637,6 +3637,15 @@ FunctionProtoType::FunctionProtoType(QualType result, ArrayRef<QualType> params,
     auto &EllipsisLoc = *getTrailingObjects<SourceLocation>();
     EllipsisLoc = epi.EllipsisLoc;
   }
+
+  if (epi.FunctionEffects) {
+    auto &ExtraBits = *getTrailingObjects<FunctionTypeExtraBitfields>();
+    ExtraBits.HasFunctionEffects = true;
+
+    // N.B. This is uninitialized storage.
+    FunctionEffectSet *PFX = getTrailingObjects<FunctionEffectSet>();
+    new (PFX) FunctionEffectSet(epi.FunctionEffects);
```

Comment:
It would also appear that neither `FunctionProtoType` nor `TrailingObjects` has any provision for invoking the destructors of the trailing objects. Is it okay that they are leaked?

Reply:
Ah yes, that is on purpose. All AST nodes (and everything stored in them) must be trivially destructible. The AST nodes are all allocated in a bump pointer allocator, so to free them, we just free that allocator. That also means that you can never put e.g. a `SmallVector` in an AST node (or anything else that has a non-trivial destructor).

Reply:
> All AST nodes must be trivially destructible

And because of that, it’s ok for us to never invoke their destructors.